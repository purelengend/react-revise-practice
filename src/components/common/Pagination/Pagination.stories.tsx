import type { Meta, StoryObj } from "@storybook/react";
import { Center } from "@chakra-ui/react";

// Components
import Pagination from ".";

const meta: Meta<typeof Pagination> = {
  title: "Components/Pagination",
  component: Pagination,
  decorators: [
    (Story) => (
      <Center style={{ width: "100vh" }}>
        <Story />
      </Center>
    ),
  ],
  argTypes: {
    totalRecords: {
      description: "The total number of records",
    },
    pageLimit: {
      description: "The number of records per page",
    },
  },
  args: {
    totalRecords: 30,
    pageLimit: 6,
  },
  parameters: {
    controls: {
      expanded: true,
    },
  },
};

export default meta;

type Story = StoryObj<typeof Pagination>;

export const Primary: Story = {};
