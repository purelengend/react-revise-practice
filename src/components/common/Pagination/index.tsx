import { QUERY_PARAMS } from "@/constants";
import { Button, Center } from "@chakra-ui/react";
import { memo, useCallback } from "react";
import { useSearchParams } from "react-router-dom";

export type PaginationProps = {
  totalRecords: number;
  pageLimit: number;
};

const Pagination = memo(({ totalRecords, pageLimit }: PaginationProps) => {
  const [searchParams, setSearchParams] = useSearchParams();

  const currentPage = Number(searchParams.get(QUERY_PARAMS.PAGE)) || 1;

  const pages: Array<number> = [];

  for (let i = 1; i <= Math.ceil(totalRecords / pageLimit); i++) {
    pages.push(i);
  }

  const handlePageChange = useCallback(
    (e: React.MouseEvent<HTMLButtonElement>) => {
      searchParams.set(QUERY_PARAMS.PAGE, e.currentTarget.value);

      setSearchParams(searchParams);
    },
    [searchParams, setSearchParams],
  );
  return (
    <Center mt={15} gap={5}>
      {pages.map((page) => {
        return (
          <Button
            key={page}
            value={page}
            px={4}
            py={1.5}
            fontSize="sm"
            bg="white.200"
            _hover={{
              bg: "yellow.100",
            }}
            _active={{
              bg: "yellow.200",
            }}
            isActive={currentPage === page}
            onClick={handlePageChange}
          >
            {page}
          </Button>
        );
      })}
    </Center>
  );
});

export default Pagination;
