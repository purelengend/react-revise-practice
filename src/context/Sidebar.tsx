import { createContext } from "react";

// Hooks
import { useSidebar } from "@/hooks";

export type SidebarContextProps = {
  sidebarState: boolean;
  openSidebar: () => void;
  closeSidebar: () => void;
};

export const SidebarContext = createContext<SidebarContextProps>(
  {} as SidebarContextProps,
);

export const SidebarContextProvider = (props: {
  children: React.ReactNode;
}) => {
  const { sidebarState, openSidebar, closeSidebar } = useSidebar();
  return (
    <SidebarContext.Provider
      value={{
        sidebarState,
        openSidebar,
        closeSidebar,
      }}
    >
      {props.children}
    </SidebarContext.Provider>
  );
};
