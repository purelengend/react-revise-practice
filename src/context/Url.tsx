import { ReactNode, createContext, useMemo } from "react";

// Hooks
import { useUrl } from "@/hooks";

export type UrlContextProps = {
  path: string;
  sortValue: string;
  setSortValue: (sortValue: string) => void;
  setOrderValue: (orderValue: string) => void;
  currentPage: number;
  setPageValue: (pageValue: number) => void;
  resetPageValue: () => void;
};

export const UrlContext = createContext<UrlContextProps>({} as UrlContextProps);

export const UrlContextProvider = (props: { children: ReactNode }) => {
  const {
    path,
    setOrderValue,
    sortValue,
    setSortValue,
    currentPage,
    setPageValue,
    resetPageValue,
  } = useUrl();

  const urlContextValue = useMemo(
    () => ({
      path,
      setOrderValue,
      sortValue,
      setSortValue,
      currentPage,
      setPageValue,
      resetPageValue,
    }),
    [
      currentPage,
      path,
      resetPageValue,
      setOrderValue,
      setPageValue,
      setSortValue,
      sortValue,
    ],
  );
  return (
    <UrlContext.Provider value={urlContextValue}>
      {props.children}
    </UrlContext.Provider>
  );
};
