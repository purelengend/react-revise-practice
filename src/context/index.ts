export * from "./Sidebar";
export * from "./Url";
