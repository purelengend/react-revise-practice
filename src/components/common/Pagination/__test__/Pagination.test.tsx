import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

// Component
import Pagination, { PaginationProps } from "..";

// Context
import { UrlContextProvider } from "@/context";

describe("Pagination test cases", () => {
  const mockPaginationProps: PaginationProps = {
    totalRecords: 30,
    pageLimit: 6,
  };

  const setup = () => {
    userEvent.setup();
    return render(
      <UrlContextProvider>
        <Pagination {...mockPaginationProps} />
      </UrlContextProvider>,
    );
  };

  it("should render correctly", () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });

  it("should invoke handlePageChange function when clicking page number", async () => {
    setup();

    const pageNumberBtn = screen.getByRole("button", {
      name: /2/i,
    });

    await userEvent.click(pageNumberBtn);
  });
});
