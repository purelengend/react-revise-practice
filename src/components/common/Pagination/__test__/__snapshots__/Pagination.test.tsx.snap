// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination test cases should render correctly 1`] = `
<div>
  <div
    class="css-vxgoms"
  >
    <button
      class="chakra-button css-1vj0tun"
      data-active=""
      type="button"
      value="1"
    >
      1
    </button>
    <button
      class="chakra-button css-1vj0tun"
      type="button"
      value="2"
    >
      2
    </button>
    <button
      class="chakra-button css-1vj0tun"
      type="button"
      value="3"
    >
      3
    </button>
    <button
      class="chakra-button css-1vj0tun"
      type="button"
      value="4"
    >
      4
    </button>
    <button
      class="chakra-button css-1vj0tun"
      type="button"
      value="5"
    >
      5
    </button>
  </div>
</div>
`;
